File: sdk/ruby/lib/anyvali/schemas/nullable_schema.rb
# frozen_string_literal: true

module AnyVali
  class NullableSchema < Schema
    attr_reader :inner_schema

    def initialize(schema:, **kwargs)
      @inner_schema = schema
      super(kind: "nullable", **kwargs)
    end

    def to_node
      node = super
      node["schema"] = @inner_schema.to_node
      node
    end

    protected

    def validate(value, path, issues, context)
      return if value.nil?
      result = @inner_schema.safe_parse(value, path: path, context: context)
      issues.concat(result.issues) if result.failure?
    end

    def dup_with(**overrides)
      schema = overrides.delete(:schema) || @inner_schema
      attrs = {
        schema: schema,
        kind: @kind,
        constraints: @constraints,
        coerce_config: @coerce_config,
        default_value: @default_value,
        has_default: @has_default,
        custom_validators: @custom_validators
      }.merge(overrides)
      self.class.new(**attrs)
    end
  end
end

File: sdk/csharp/src/AnyVali/Schemas/NullableSchema.cs
namespace AnyVali.Schemas;

public sealed class NullableSchema : Schema<object?>
{
    internal Schema Inner { get; }

    public NullableSchema(Schema inner)
    {
        Inner = inner;
    }

    public new NullableSchema Default(object? value) => (NullableSchema)base.Default(value);

    internal override object? Validate(object? input, ValidationContext ctx)
    {
        if (input is null) return null;
        return Inner.Validate(input, ctx);
    }

    internal override object? RunPipeline(object? input, ValidationContext ctx)
    {
        if (input is null) return null;
        return Inner.RunPipeline(input, ctx);
    }

    internal override Dictionary<string, object?> ToNode()
    {
        var node = new Dictionary<string, object?>
        {
            ["kind"] = "nullable",
            ["inner"] = Inner.ToNode(),
        };
        AddDefaultAndCoercion(node);
        return node;
    }

    internal override Schema Clone() => new NullableSchema(Inner)
    {
        DefaultValue = DefaultValue, CoercionCfg = CoercionCfg, IsPortable = IsPortable,
    };
}
